pkg/sigv4/sigv4.go

```diff
@@ -0,0 +1,200 @@
+package sigv4
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+)
+
+const signingAlgo = "AWS4-HMAC-SHA256"
+const awsServiceName = "aps"
+
+type Signer interface {
+	Sign(req *http.Request) error
+}
+
+type DefaultSigner struct {
+	iSO8601Date      string
+	canonicalHeaders string
+	signedHeaders    string
+	credentialScope  string
+	config           *Config
+	payloadHash      string
+}
+
+func NewDefaultSigner(config *Config) Signer {
+	return &DefaultSigner{
+		config: config,
+	}
+}
+
+func (d *DefaultSigner) Sign(req *http.Request) error {
+	now := time.Now().UTC()
+	d.iSO8601Date = now.Format("20060102T150405Z")
+	d.credentialScope = fmt.Sprintf(
+		"%s/%s/%s/aws4_request",
+		now.UTC().Format("20060102"),
+		d.config.Region,
+		awsServiceName,
+	)
+
+	payloadHash, err := d.getPayloadHash(req)
+	if err != nil {
+		return err
+	}
+
+	d.payloadHash = payloadHash
+	d.addRequiredHeaders(req)
+	d.canonicalHeaders, d.signedHeaders = d.getCanonicalAndSignedHeaders(req)
+
+	canonicalReq := d.createCanonicalRequest(req)
+	stringToSign, err := d.createStringToSign(canonicalReq)
+	if err != nil {
+		return err
+	}
+
+	signature := d.sign(d.createSigningKey(), stringToSign)
+	authorizationHeader := fmt.Sprintf(
+		"%s Credential=%s/%s, SignedHeaders=%s, Signature=%s",
+		signingAlgo,
+		d.config.AwsAccessKeyID,
+		d.credentialScope,
+		d.signedHeaders,
+		signature,
+	)
+	req.Header.Set("Authorization", authorizationHeader)
+	return nil
+}
+
+func (d *DefaultSigner) getPayloadHash(req *http.Request) (string, error) {
+	if req.Body == nil {
+		return hex.EncodeToString(sha256.New().Sum(nil)), nil
+	}
+
+	reqBody, err := io.ReadAll(req.Body)
+	if err != nil {
+		return "", err
+	}
+	reqBodyBuffer := bytes.NewReader(reqBody)
+
+	hash := sha256.New()
+	if _, err := io.Copy(hash, reqBodyBuffer); err != nil {
+		return "", err
+	}
+
+	payloadHash := hex.EncodeToString(hash.Sum(nil))
+
+	// ensuring that we keep the request body intact for next tripper
+	req.Body = io.NopCloser(bytes.NewReader(reqBody))
+
+	return payloadHash, nil
+}
+
+func (d *DefaultSigner) addRequiredHeaders(req *http.Request) {
+	req.Header.Set("Host", req.Host)
+	req.Header.Set("x-amz-date", d.iSO8601Date)
+	req.Header.Set("x-amz-content-sha256", d.payloadHash)
+}
+
+func (d *DefaultSigner) getCanonicalAndSignedHeaders(req *http.Request) (string, string) {
+	var headers []string
+	var signedHeaders []string
+
+	for key, value := range req.Header {
```

Comment:
I'm checking the [AWS SDK implementation](https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/v4/v4.go#L434-L436), and it has a bit more complicated logic, like skipping from signing some keys, do you think we need something similar here?

The same probably applies to the query params

Reply:
addressed in [f3e0598](https://github.com/grafana/xk6-output-prometheus-remote/pull/169/commits/f3e0598fb8897a0600f8f5030b8696c25e1553d0)